lib/resty/socket.lua

```diff
@@ -22,6 +24,16 @@ do
   end
 
   proxy_mt = {
+    connect = function(self, ...)
+      if self._connected_before then
+        -- luasocket does not allow reusing sockets after being closed
+        local socket = require 'socket'
+        self.sock = socket.tcp()
+      else
+        self._connected_before = true
+      end
+      return self.sock:connect(...)
+    end,
```

Comment:
This patch has to somehow prevent calling `connect()` twice, or else it seems to me like doing so will simply replace the socket; what happens to the established connection and existing socket? Does the connection RST? If so, that's no good.

Reply:
I tried initially, because the socket gave me a "closed" error. So I set a flag on the `close` and `keepalive` methods, and only renewed the socket when connect was called and that flag set. But that doesn't work when the remote closes the connection. And then we still get the "closed" error upon calling `connect` again. So this is the best I could come up with.

At worst it obfuscates the problem, if in the middle of a connection you call connect again, that should always be an error (resty and luasocket). So in this case the behaviour still does something weird, replacing the socket, which will cause other application errors probably and hence obfuscate the real issue. That said, this only happens in an already erroneous situation. While it fixes the good-path of reusing sockets after they are closed.

The only way around it is probably having a fail-safe way of detecting the socket to be closed

Reply:
@thibaultcha did the patch below resolve this one? or do you have concerns left?

Reply:
I rebased the PR, patch went in to the original commit, I'll resolve this, please have another look